Switch internal array type to contain optional values

tomijaga opened the pull request.

tomijaga committed:
Switch internal array type to contain optional values

ByronBecker reviewed (commented):
Hi @tomijaga, thanks for your interest and contribution!

I'm not sure why the buffer explicitly needs to contain optional values. Can you give me a use case?

You can do something like this already. 

```
do {
  let buf = B.init<?Nat>();
  B.add(buf, ?2);
  B.add(buf, null);
  B.add(buf, ?4);
  B.add(buf, ?6);

  assert (B.toArray() == [?2, null, ?4, ?6]);
}
```

Also, I don't recommend using the `unwrap` method you have as written that uses the internal `Prelude.unreachable` trap method. It forces a canister to trap instead of handling nulls programmatically via a switch statement, and is generally [not recommended by the Motoko team](https://github.com/dfinity/motoko-base/blob/c174fc4bfc0287ccbfbd5a4b55edbc8669e3f2f3/src/Option.mo#L154).

tomijaga commented:
Hey @ByronBecker, thank you for the feedback. I clicked the create PR button thinking it would make one for my forked repo instead of the main repo. 

 I am creating a [BitBuffer](https://github.com/NatLabs/BitBuffer) and [BitSet](https://github.com/NatLabs/BitSet) in Motoko, which use the StableBuffer internally. I updated your implementation to find out if the StableBuffer would use less memory if the extra space in the `elems` array was filled with `null` values. 
 I tested it using the [motoko playground](https://m7sm4-2iaaa-aaaab-qabra-cai.ic0.app/?tag=1477932713
) and it turns out that my version is less efficient. My suspicion is that the `Option` type is similar to a union struct where the memory allocated is the max memory of all its members.  

Sorry to bother you with this PR. I will go ahead and close it.

Also, is it okay if I message you on the dfinity forum regarding my BitBuffer and BitSet implementation? I would love to hear your thoughts and ways to improve it.

ByronBecker commented:
@tomijaga, of course, happy to answer questions or provide any help that I can. Feel free to message me on the forums - my username there is [icme](https://forum.dfinity.org/u/icme/summary).

tomijaga commented:
Hey @ByronBecker, it's been a while since I submitted this PR. 
I wanted to discuss my reasoning behind it even though my tests didn't bring the results I had hoped. 
I tested for the number of wasm instructions called using the `performanceCount()` fn instead of comparing the memory used. 
I am writing again because the base repo has switched its implementation to using [optional values internally](https://github.com/dfinity/motoko-base/blob/d52aecd7dabe3f5f8169b7680fe0a0702a67b0e3/src/Buffer.mo#L62). 

Currently, when initializing a buffer, the internal array is not created until the user calls the adds the first element. The element is then copied to every single position in the array to fill it. The unnecessary duplication here wastes memory, especially if the values are compound data types. 
Using `null` values to fill the array positions where elements have not yet been added saves space. 

I've removed the `unwrap()` method and replaced it with a switch method in `get()`. 
I'm looking forward to hearing your thoughts on the PR.

tomijaga committed:
remove unwrap

ByronBecker commented on src/StableBuffer.mo:
```diff
@@ -86,88 +89,95 @@ module {
+  Prim.Array_tabulate<X>(
+    buffer.count,
+    func(x : Nat) : X {
+      get(buffer, x);
```
All of these changes to use `get` add an extra function call (small amount of overhead). Not a huge deal, but directly using the `buffer.elems[index]` isn't that much cognitive overload.

Just thought I'd mention in case you have a strong opinion on this.

tomijaga replied:
The reason for using `get()` in this function is to remove the option type. We would still have to use the `switch` statement if we use `buffer.elems[index]` directly and duplicate the `Prim.trap` message.

ByronBecker commented on src/StableBuffer.mo:
```diff
@@ -86,88 +89,95 @@ module {
+  public func get<X>(buffer : StableBuffer<X>, i : Nat) : X {
+    switch (buffer.elems[i]) {
+      case (?elem) elem;
+      case (null) Prim.trap("Buffer index out of bounds in get");
```
One of the issues with moving this to `optional` type land is that now a developer can insert a few items, then set index 0 to `null`, and then this would trap on `buffer.get(0)`. I guess you can already mess around with the mutable `count` field now though 🤔 

The "new" implementation in base is a class based implementation (requires transformation in order to achieve stability), but has the benefit of the elements and count being private instance variables that can't be modified by the developer.

I'm sort of on the fence about just ripping all the new Buffer implementation out of the class and into functions (just like I did for the old Buffer implementation), but that new module is just so huge 😆 

Another thing to consider is that this would be a breaking change in terms of the underlying data structure, so I'll need to bump the major version (not a huge deal).

tomijaga replied:
I hadn't thought of that 😲. It's hard to tamper-proof the internal state. I think the alternative would be to switch to a functional implementation that returns a new state after every function.

I was wondering if this same issue could happen if the value set by the user is also an optional type, making the internal type a nested option (`??X`). But after testing this in the [Motoko playground](https://m7sm4-2iaaa-aaaab-qabra-cai.raw.ic0.app/?tag=2329101299) it works quite well.

Yes, I agree that the functions included in the `Buffer` module are a lot and don't need to be added to this one. I would only like to add the essential function in the `Buffer` class like `insert` and `remove`. 

I don't have a lot of experience publishing libs for users. So I don't have a developed sense of what's considered a breaking or a minor change. Why is it a breaking change in this case? When to the user, none of the functions really change.

ByronBecker replied:
> "I don't have a lot of experience publishing libs for users. So I don't have a developed sense of what's considered a breaking or a minor change. Why is it a breaking change in this case? When to the user, none of the functions really change."

So even though the interfaces are the same, the underlying types are different.

For example, the [current Buffer class](https://github.com/dfinity/motoko-base/blob/aafcdee0c8328087aeed506e64aa2ff4ed329b47/src/Buffer.mo#L67) is a class and therefore is **not** stable (requires preupgrade/postupgrade).

Also, the class-based implementation internals (elems, count) are private, so they are not directly accessible by developers. If I'm using a class-based implementation, I know that by changing the types, there's no way that end users are currently directly pulling from and using the data structure internals in their code, but are going through the class-based interfaces.

If I make this change from [var X] to [var ?X], then this means developers who have an existing buffer

`stable let myBuffer: StableBuffer<X> = ...`

which has underlying type

```
public type StableBuffer<X> = {
  initCapacity : Nat;
  var count : Nat;
  var elems : [var ?X];
};
```

Will now become underlying type

```
public type StableBuffer<X> = {
  initCapacity : Nat;
  var count : Nat;
  var elems : [var ?X];
};
```

So you'll end up receiving an error when you attempt to re-deploy. Something like:

```
WARNING!
Stable interface compatibility check failed for canister 'example'.
Upgrade will either FAIL or LOSE some stable variable data.

(unknown location): Compatibility error [M0170], stable variable myBuffer of previous type
  {var count : Nat; var elems : [var Nat]; initCapacity : Nat}
cannot be consumed at new type
  {var count : Nat; var elems : [var ?Nat]; initCapacity : Nat}

Do you want to proceed? yes/No
```

Check out the [migration testing I did](https://github.com/canscale/StableBuffer/commit/f4a8c13aa263687aebe691565016c426048d5582), following these steps.

1. Navigate to the `example` folder and `dfx deploy` the example.
2. Enter a few values into the buffer via `addNatToBuffer()` and test out that they were inserted
3. Sub in the new stable buffer
  a. Comment out the old version (first half of `src/StableBuffer.mo` file Lines 1-174), and uncomment the new StableBuffer.mo implementation (second half of src/StableBuffer.mo Lines 180-end)
4. Try to redeploy - you'll notice the breaking change in the example/src/main.mo actor under the `printBuffer()` function.
5. Sub in the new version of the actor (comment out `example/src/main.mo` actor at first half of file, uncomment second actor at bottom of file)
6. Try to re-deploy. You'll get the interface compatibility warning I showed above ☝️. This will scare away many developers. 
7. Continue through with the deployment and type yes. You'll notice that despite the warning, the migration was successful (this was baffling to me, but I think it has something to do with the `postUpgrade()` deserialization)

Despite the fact that this upgrade was stable, it still is a breaking change - it broke an API (required an update to the logic) due to the fact that the buffer internals are accessible.

An idea came to me while reviewing this PR. Check out [this](https://github.com/canscale/StableBuffer/commit/f4a8c13aa263687aebe691565016c426048d5582#diff-38735e3d38cb2f65640007a8695265c9d43a5ff4dd66a1dbf2aa9f64cd6272b2). What do you think of just adding a new module called `StableOptBuffer` for your use case? This way no one gets broken, and they can migrate if they choose to rather than being forced to by the library overwriting the previous StableBuffer module.

ByronBecker commented on src/StableBuffer.mo:
```diff
@@ -86,88 +89,95 @@ module {
-      buffer.count,
-      func(x : Nat) : X { buffer.elems[x] }
-    );
+  public func toArray<X>(buffer : StableBuffer<X>) : [X] =
```
Here's another instance where if the developer sets one of the elements with an `index < count` to null, this will trap when copying it over to an array. This trap will happen at runtime instead of at compile time, and feels a bit unsafe-ish/buggy to me.

If we make these changes, then I definitely need to start including usage warnings all over the code 😅

tomijaga replied:
I tried this example in the [motoko playground](https://m7sm4-2iaaa-aaaab-qabra-cai.raw.ic0.app/?tag=2912498390). If you notice the `say()` fn doesn't trap even though you enter a `null` value. This means that Motoko supports optional null values `?null` 💪 .

ByronBecker replied:
> "This means that Motoko supports optional null values"

🤯

ByronBecker commented on src/StableBuffer.mo:
```diff
@@ -86,88 +89,95 @@ module {
 
   /// Gets the `i`-th element of the buffer as an option. Returns `null` when `i >= count`. Indexing is zero-based.
-  public func getOpt<X>(buffer: StableBuffer<X>, i : Nat) : ?X {
+  public func getOpt<X>(buffer : StableBuffer<X>, i : Nat) : ?X {
```
If the buffer is changed internally to have elems of type `?X`, then is this function still needed? (Can just use `get`)

tomijaga replied:
I think this function still serves its purpose even with the `?X` type. This function retrieves the value of any index inside or outside the buffer without trapping.

ByronBecker reviewed (commented):
Thanks for the contribution! A few comments/questions I'd like to discuss before approving this breaking change

ByronBecker commented:
Hey @tomijaga, given where the new Buffer class is right now, I'd be down to merge in this commit from the additions you've made. I think what I'll do is just bump the version to v1.0.0

https://github.com/NatLabs/StableBuffer/commit/7007932e864e3c08354e80fe603e45e05b54d2ad

If you still want to merge this work in, I'll accept this commit ☝️  if you push it to this PR

tomijaga commented:
Hey @ByronBecker, that sounds good!

However, I'm not sure I understand the request as I have already merged the https://github.com/NatLabs/StableBuffer/commit/7007932e864e3c08354e80fe603e45e05b54d2ad (`remove unwrap`) commit into this PR [earlier](https://github.com/canscale/StableBuffer/pull/4#event-7779142055). 
Did you want me to update this PR with the latest changes in `main` or merge the commit with a different PR?

ByronBecker commented:
@tomijaga Ah, I was looking at your main branch for some reason. No this looks good. going to merge this in and tag is as `v1.0.0`
